Index: includes/common.inc
===================================================================
RCS file: /cvs/drupal/drupal/includes/common.inc,v
retrieving revision 1.611.2.12
diff -u -F^f -r1.611.2.12 common.inc
--- includes/common.inc	17 Oct 2007 21:28:59 -0000	1.611.2.12
+++ includes/common.inc	4 Jan 2008 20:13:04 -0000
@@ -1469,15 +1469,15 @@ function drupal_get_css($css = NULL) {
           // If a CSS file is not to be preprocessed and it's a module CSS file, it needs to *always* appear at the *top*,
           // regardless of whether preprocessing is on or off.
           if (!$preprocess && $type == 'module') {
-            $no_module_preprocess .= '<style type="text/css" media="'. $media .'">@import "'. base_path() . $file .'";</style>' ."\n";
+            $no_module_preprocess .= '<style type="text/css" media="'. $media .'">@import "'. cdn_file($file) .'";</style>' ."\n";
           }
           // If a CSS file is not to be preprocessed and it's a theme CSS file, it needs to *always* appear at the *bottom*,
           // regardless of whether preprocessing is on or off.
           else if (!$preprocess && $type == 'theme') {
-            $no_theme_preprocess .= '<style type="text/css" media="'. $media .'">@import "'. base_path() . $file .'";</style>' ."\n";
+            $no_theme_preprocess .= '<style type="text/css" media="'. $media .'">@import "'. cdn_file($file) .'";</style>' ."\n";
           }
           else {
-            $output .= '<style type="text/css" media="'. $media .'">@import "'. base_path() . $file .'";</style>' ."\n";
+            $output .= '<style type="text/css" media="'. $media .'">@import "'. cdn_file($file) .'";</style>' ."\n";
           }
         }
       }
@@ -1486,7 +1486,7 @@ function drupal_get_css($css = NULL) {
     if ($is_writable && $preprocess_css) {
       $filename = md5(serialize($types)) .'.css';
       $preprocess_file = drupal_build_css_cache($types, $filename);
-      $output .= '<style type="text/css" media="'. $media .'">@import "'. base_path() . $preprocess_file .'";</style>'. "\n";
+      $output .= '<style type="text/css" media="'. $media .'">@import "'. cdn_file($file) .'";</style>'. "\n";
     }
   }
 
@@ -1680,7 +1680,7 @@ function drupal_get_js($scope = 'header'
         break;
       default:
         foreach ($data as $path => $info) {
-          $output .= '<script type="text/javascript"'. ($info['defer'] ? ' defer="defer"' : '') .' src="'. check_url(base_path() . $path) . ($info['cache'] ? '' : '?'. time()) ."\"></script>\n";
+          $output .= '<script type="text/javascript"'. ($info['defer'] ? ' defer="defer"' : '') .' src="'. check_url(cdn_file($path)) . ($info['cache'] ? '' : '?'. time()) ."\"></script>\n";
         }
     }
   }
Index: includes/file.inc
===================================================================
RCS file: /cvs/drupal/drupal/includes/file.inc,v
retrieving revision 1.90.2.1
diff -u -F^f -r1.90.2.1 file.inc
--- includes/file.inc	31 May 2007 05:48:58 -0000	1.90.2.1
+++ includes/file.inc	4 Jan 2008 20:13:06 -0000
@@ -33,7 +33,7 @@ function file_create_url($path) {
   }
   switch (variable_get('file_downloads', FILE_DOWNLOADS_PUBLIC)) {
     case FILE_DOWNLOADS_PUBLIC:
-      return $GLOBALS['base_url'] .'/'. file_directory_path() .'/'. str_replace('\\', '/', $path);
+      return cdn_file(file_directory_path() .'/'. str_replace('\\', '/', $path));
     case FILE_DOWNLOADS_PRIVATE:
       return url('system/files/'. $path, NULL, NULL, TRUE);
   }
Index: includes/theme.inc
===================================================================
RCS file: /cvs/drupal/drupal/includes/theme.inc,v
retrieving revision 1.337.2.2
diff -u -F^f -r1.337.2.2 theme.inc
--- includes/theme.inc	31 May 2007 05:52:42 -0000	1.337.2.2
+++ includes/theme.inc	4 Jan 2008 20:13:11 -0000
@@ -322,24 +322,24 @@ function theme_get_setting($setting_name
 
     if ($settings['toggle_logo']) {
       if ($settings['default_logo']) {
-        $settings['logo'] = base_path() . dirname($theme_object->filename) .'/logo.png';
+        $settings['logo'] = cdn_file(dirname($theme_object->filename) .'/logo.png');
       }
       elseif ($settings['logo_path']) {
-        $settings['logo'] = base_path() . $settings['logo_path'];
+        $settings['logo'] = $base . $settings['logo_path'];
       }
     }
 
     if ($settings['toggle_favicon']) {
       if ($settings['default_favicon']) {
         if (file_exists($favicon = dirname($theme_object->filename) .'/favicon.ico')) {
-          $settings['favicon'] = base_path() . $favicon;
+          $settings['favicon'] = $base . $favicon;
         }
         else {
-          $settings['favicon'] = base_path() . 'misc/favicon.ico';
+          $settings['favicon'] = $base . 'misc/favicon.ico';
         }
       }
       elseif ($settings['favicon_path']) {
-        $settings['favicon'] = base_path() . $settings['favicon_path'];
+        $settings['favicon'] = $base . $settings['favicon_path'];
       }
       else {
         $settings['toggle_favicon'] = FALSE;
@@ -613,8 +613,7 @@ function theme_links($links, $attributes
 function theme_image($path, $alt = '', $title = '', $attributes = NULL, $getsize = TRUE) {
   if (!$getsize || (is_file($path) && (list($width, $height, $type, $image_attributes) = @getimagesize($path)))) {
     $attributes = drupal_attributes($attributes);
-    $url = (url($path) == $path) ? $path : (base_path() . $path);
-    return '<img src="'. check_url($url) .'" alt="'. check_plain($alt) .'" title="'. check_plain($title) .'" '. $image_attributes . $attributes .' />';
+    return '<img src="'. check_url(cdn_file(url($path))) .'" alt="'. check_plain($alt) .'" title="'. check_plain($title) .'" '. $image_attributes . $attributes .' />';
   }
 }
 
Index: themes/garland/page.tpl.php
===================================================================
RCS file: /cvs/drupal/drupal/themes/garland/page.tpl.php,v
retrieving revision 1.5
diff -u -F^f -r1.5 page.tpl.php
--- themes/garland/page.tpl.php	14 Dec 2006 00:37:00 -0000	1.5
+++ themes/garland/page.tpl.php	4 Jan 2008 20:13:36 -0000
@@ -6,9 +6,9 @@
     <?php print $head ?>
     <?php print $styles ?>
     <?php print $scripts ?>
-    <style type="text/css" media="print">@import "<?php print base_path() . path_to_theme() ?>/print.css";</style>
+    <style type="text/css" media="print">@import "<?php print $static_base . path_to_theme() ?>/print.css";</style>
     <!--[if lt IE 7]>
-    <style type="text/css" media="all">@import "<?php print base_path() . path_to_theme() ?>/fix-ie.css";</style>
+    <style type="text/css" media="all">@import "<?php print $static_base . path_to_theme() ?>/fix-ie.css";</style>
     <![endif]-->
   </head>
   <body<?php print phptemplate_body_class($sidebar_left, $sidebar_right); ?>>
